Geometry_Engine/Query/Product.cs

```diff
@@ -32,28 +32,28 @@ public static partial class Query
 
         public static double DotProduct(this Vector a, Vector b)
         {
-            return (a.X * b.X + a.Y * b.Y + a.Z * b.Z);
+            return a != null && b != null ? (a.X * b.X + a.Y * b.Y + a.Z * b.Z) : double.NaN;
```

Comment:
Good to add an error message here if any of the vectors are null I think.

Reply:
I skipped it for these just to because I didn't want to add a bunch of complexity to such simple methods, but you're probably right that these methods deserve some error messages too.